dali/kernels/signal/window/extract_windows_gpu.cuh

```diff
@@ -0,0 +1,568 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SIGNAL_WINDOW_EXTRACT_WINDOWS_GPU_CUH_
+#define DALI_KERNELS_SIGNAL_WINDOW_EXTRACT_WINDOWS_GPU_CUH_
+
+#include <cuda_runtime.h>
+#include <algorithm>
+
+#include "dali/core/convert.h"
+#include "dali/core/boundary.h"
+#include "dali/kernels/signal/window/extract_windows_args.h"
+#include "dali/kernels/kernel.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+namespace window {
+
+constexpr int kBlock = 32;
+
+
+/// @brief Extract and store kBlock values from kBlock windows.
+///
+/// This function reads kBlock elements from kBlock windows to shared memory
+/// and stores the transposed result to output in columns.
+///
+/// @remarks This function must be executed by all (or no) threads in a block!
+template <int num_pages = 1, typename Dst, typename Src>
+__device__ void ExtractVerticalWindowsBlock(
+    int first_window_idx,
+    Dst *__restrict__ dst, ptrdiff_t num_windows, ptrdiff_t stride,
+    const Src *__restrict__ src, ptrdiff_t length,
+    const float *__restrict__ window,
+    int win_len,
+    int win_center,
+    int step,
+    bool reflect,
+    int page = 0) {
+  __shared__ float tmp[num_pages][kBlock][kBlock+1];  // +1 to avoid bank conflicts
+  ptrdiff_t in_window_idx = first_window_idx + threadIdx.y;
```

Comment:
I don't know if it is worth to add some check if `num_pages > page`.

Reply:
I don't think it's worth it. It's only used twice and not intended for general use.